squirtgun-platform-discord/src/main/java/net/lucypoulton/squirtgun/discord/DiscordCommandListener.java

```
@@ -0,0 +1,81 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package net.lucypoulton.squirtgun.discord;
+
+import net.dv8tion.jda.api.entities.TextChannel;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import net.kyori.adventure.text.Component;
+import net.lucypoulton.squirtgun.command.context.StringContext;
+import net.lucypoulton.squirtgun.command.node.CommandNode;
+import net.lucypoulton.squirtgun.format.FormatProvider;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Predicate;
+
+public class DiscordCommandListener extends ListenerAdapter {
+
+    private final DiscordPlatform platform;
+    private final String prefix;
+    private final boolean ignoreBots;
+    private final Predicate<TextChannel> canAccept;
+
+    private final Map<String, CommandNode<?>> nodes = new HashMap<>();
+    private final Map<String, FormatProvider> formatProviders = new HashMap<>();
+
+    public DiscordCommandListener(DiscordPlatform platform, String prefix, boolean ignoreBots, Predicate<TextChannel> canAccept) {
+        this.platform = platform;
+        this.prefix = prefix;
+        this.ignoreBots = ignoreBots;
+        this.canAccept = canAccept;
+        platform.jda().addEventListener(this);
+    }
+
+    public void registerCommand(CommandNode<?> node, FormatProvider formatProvider) {
+        if (nodes.containsKey(node.getName())) {
+            throw new IllegalArgumentException("Node with given name already exists!");
+        }
+        nodes.put(node.getName(), node);
+        formatProviders.put(node.getName(), formatProvider);
+    }
```

Do we really want to throw an exception? Or maybe use something like `Map::putIfAbsent` and ignore silently? Or perhaps override the existing command in the map entirely?

---

overriding is probably the best option, i'll update that now